src/os/bluestore/BlueStore.cc

```diff
@@ -13432,7 +13432,8 @@ int BlueStore::_do_alloc_write(
 
       // FIXME: memory alignment here is bad
       bufferlist t;
-      int r = c->compress(wi.bl, t);
+      boost::optional<int32_t> compressor_message(0);
```

Comment:
Now assigning to zero is doubtful. Imagine that later one needs to implement the same parameter passing for another compressor similarly as you're doing for zlib now. Originally this hypothetical compressor applied some constant (-N) internally and now starts to vary and expose this parameter. Those originally compressed chunks should be decompressed with -N but external compression header (of version 2 having dummy compressor_message param encoded) will provide compressor message set to 0. IMO this might cause some decompression issues which are pretty hard to troubleshoot.
I think now the better approach would be to leave compressor_message in <not-set> state (which is determined in contrary to simple "int32_t compressor message;" declaration).  And do conditional encoding/decoding for headers, i.e. header decoding to use bool flag to determine if compressor_message is present:
decode()
...
  bool cmsg_presnt;
  decode(cmsg_present, p);
  if(cmsg_present)
    decode(v.compressor_message)
}
and symmetric encode which verifies if compressor_message is set and encodes bool flag and optionally the message.